go-controller/pkg/node/gateway_shared_intf.go

```diff
@@ -854,19 +872,24 @@ func (npw *nodePortWatcher) DeleteEndpointSlice(epSlice *discovery.EndpointSlice
 
 	klog.V(5).Infof("Deleting endpointslice %s in namespace %s", epSlice.Name, epSlice.Namespace)
 	// remove rules for endpoints and add back normal ones
-	namespacedName, err := serviceNamespacedNameFromEndpointSlice(epSlice)
+	namespacedName, err := util.ServiceNamespacedNameFromEndpointSlice(epSlice)
 	if err != nil {
 		return fmt.Errorf("cannot delete %s/%s from nodePortWatcher: %v", epSlice.Namespace, epSlice.Name, err)
 	}
 	epSlices, err := npw.watchFactory.GetEndpointSlices(epSlice.Namespace, epSlice.Labels[discovery.LabelServiceName])
 	if err != nil {
-		return fmt.Errorf("could not fetch endpointslices for service %s during endpointSliceDelete: %v", namespacedName, err)
+		if !kerrors.IsNotFound(err) {
+			return fmt.Errorf("error retrieving all endpointslices for service %s/%s during endpointslice delete on %s: %w",
+				namespacedName.Namespace, namespacedName.Name, epSlice.Name, err)
+		}
+		klog.V(5).Infof("No endpointslices found for service %s/%s during endpointslice delete on %s: %v",
+			namespacedName.Namespace, namespacedName.Name, epSlice.Name, err)
 	}
 
 	svc, err := npw.watchFactory.GetService(namespacedName.Namespace, namespacedName.Name)
 	if err != nil && !kerrors.IsNotFound(err) {
-		return fmt.Errorf("error while retrieving service for endpointslice %s/%s during delete: %v",
-			epSlice.Namespace, epSlice.Name, err)
+		return fmt.Errorf("error retrieving service %s/%s for endpointslice %s during endpointslice delete: %v",
```

Comment:
For the reasons in my other comment (https://github.com/ovn-org/ovn-kubernetes/pull/3485/files#r1150543228) and also for the fact that we apply the retry logic also here in ovnk node, the service might be gone already when we are executing these lines.

So it makes more sense to lookup the cached `serviceConfig` instead of looking up the service in the informer cache. 
What do you think of the following?

```
	epSlices, err := npw.watchFactory.GetEndpointSlices(epSlice.Namespace, epSlice.Labels[discovery.LabelServiceName])
	if err != nil {
		if !kerrors.IsNotFound(err) {
			return fmt.Errorf("error retrieving all endpointslices for service %s/%s during endpointslice delete on %s: %w",
				namespacedName.Namespace, namespacedName.Name, epSlice.Name, err)
		}
		klog.V(5).Infof("No other endpointslices found for service %s/%s during endpointslice delete on %s: %v",
			namespacedName.Namespace, namespacedName.Name, epSlice.Name, err)
		epSlices = []*discovery.EndpointSlice{epSlice}
	}

	if oldSvcConfig, exists := npw.getServiceInfo(namespacedName); exists {
		// Lock the cache mutex here so we don't miss a service delete during an endpoint delete
		// we have to do this because deleting and adding iptables rules is slow.
		npw.serviceInfoLock.Lock()
		defer npw.serviceInfoLock.Unlock()

		localEndpoints := npw.GetLocalEndpointAddresses(epSlices, oldSvcConfig.service)
		svcConfig, _ := npw.updateServiceInfo(namespacedName, nil, &hasLocalHostNetworkEp, localEndpoints)

		if err = delServiceRules(svcConfig.service, svcConfig.localEndpoints.UnsortedList(), npw); err != nil {
			errors = append(errors, err)
		}
		if err = addServiceRules(svcConfig.service, localEndpoints.UnsortedList(), hasLocalHostNetworkEp, npw); err != nil {
			errors = append(errors, err)
		}
		return apierrors.NewAggregate(errors)
	}
	return nil
```

Reply:
I'm actually going to keep the current code here. No need to over-complicate the logic when the current one already handles the case I was describing.